core/src/main/java/org/infinispan/container/AbstractSegmentedDataContainer.java

```diff
@@ -0,0 +1,634 @@
+   @Inject protected Cache<K, V> cache;
+   protected KeyPartitioner keyPartitioner;
+
+   protected final List<Consumer<Iterable<InternalCacheEntry<K, V>>>> listeners = new CopyOnWriteArrayList<>();
+
+   protected abstract ConcurrentMap<K, InternalCacheEntry<K, V>> getMapForSegment(int segment);
+   protected abstract int getSegmentForKey(Object key);
+
+   protected abstract Log log();
+   protected abstract boolean trace();
+
+   @Start
+   public void start() {
+      HashConfiguration hashConfiguration = cache.getCacheConfiguration().clustering().hash();
+      keyPartitioner = hashConfiguration.keyPartitioner();
+   }
+
+   @Override
+   public InternalCacheEntry<K, V> get(int segment, Object k) {
+      ConcurrentMap<K, InternalCacheEntry<K, V>> map = getMapForSegment(segment);
+      InternalCacheEntry<K, V> e = map != null ? map.get(k) : null;
+      if (e != null && e.canExpire()) {
+         long currentTimeMillis = timeService.wallClockTime();
+         if (e.isExpired(currentTimeMillis) &&
+               expirationManager.entryExpiredInMemory(e, currentTimeMillis).join() == Boolean.TRUE) {
+            e = null;
+         } else {
+            e.touch(currentTimeMillis);
+         }
+      }
+      return e;
+   }
+
+   @Override
+   public InternalCacheEntry<K, V> get(Object k) {
+      return get(getSegmentForKey(k), k);
+   }
+
+   @Override
+   public InternalCacheEntry<K, V> peek(int segment, Object k) {
+      ConcurrentMap<K, InternalCacheEntry<K, V>> entries = getMapForSegment(segment);
+      if (entries != null) {
+         if (entries instanceof PeekableMap) {
+            return ((PeekableMap<K, InternalCacheEntry<K, V>>) entries).peek(k);
+         }
+         return entries.get(k);
+      }
+      return null;
+   }
+
+   @Override
+   public InternalCacheEntry<K, V> peek(Object k) {
+      return peek(getSegmentForKey(k), k);
+   }
+
+   @Override
+   public void put(int segment, K k, V v, Metadata metadata) {
+      ConcurrentMap<K, InternalCacheEntry<K, V>> entries = getMapForSegment(segment);
+      if (entries != null) {
+         boolean l1Entry = false;
+         if (metadata instanceof L1Metadata) {
+            metadata = ((L1Metadata) metadata).metadata();
+            l1Entry = true;
+         }
+         InternalCacheEntry<K, V> e = entries.get(k);
+
+         if (trace()) {
+            log().tracef("Creating new ICE for writing. Existing=%s, metadata=%s, new value=%s", e, metadata, toStr(v));
+         }
+         final InternalCacheEntry<K, V> copy;
+         if (l1Entry) {
+            copy = entryFactory.createL1(k, v, metadata);
+         } else if (e != null) {
+            copy = entryFactory.update(e, v, metadata);
+         } else {
+            // this is a brand-new entry
+            copy = entryFactory.create(k, v, metadata);
+         }
+
+         if (trace())
+            log().tracef("Store %s in container", copy);
+
+         entries.compute(copy.getKey(), (key, entry) -> {
```

Comment:
Hrmm probably, this was just copied from the `DefaultDataContainer`. I will try to ensure we aren't calling the trace method inside the `compute` method. This can be done afterwards though, as I don't want to toy with this right now.